ipld/read.go

```diff
@@ -18,190 +18,162 @@ import (
 func RetrieveData(
-	ctx context.Context,
+	parentCtx context.Context,
 	dah *da.DataAvailabilityHeader,
-	dag ipld.NodeGetter,
+	dag ipld.DAGService,
 	codec rsmt2d.Codec,
 ) (*rsmt2d.ExtendedDataSquare, error) {
 	edsWidth := len(dah.RowsRoots)
-	sc := newshareCounter(ctx, uint32(edsWidth))
-	// convert the row and col roots into Cids
 	rowRoots := dah.RowsRoots
-	colRoots := dah.ColumnRoots
-	// sample 1/4 of the total extended square by sampling half of the leaves in
-	// half of the rows.
-	// Sampling is done randomly here in an effort to increase
-	// resilience when the downloading node is also serving the shares back to
-	// the network. // TODO @renaynay: this should probably eventually change to
-	// just downloading the first half.
-	for _, row := range uniqueRandNumbers(edsWidth/2, edsWidth) {
-		for _, col := range uniqueRandNumbers(edsWidth/2, edsWidth) {
-			rootCid, err := plugin.CidFromNamespacedSha256(rowRoots[row])
-			if err != nil {
-				return nil, err
-			}
+	dataSquare := make([][]byte, edsWidth*edsWidth)
 
-			go sc.retrieveShare(rootCid, true, row, col, dag)
-		}
+	quadrant := pickRandomQuadrant(rowRoots)
+	errGroup, ctx := errgroup.WithContext(parentCtx)
+	errGroup.Go(func() error {
+		return fillQuadrant(ctx, quadrant, dag, true, dataSquare)
+	})
+	if err := errGroup.Wait(); err != nil {
+		return nil, err
 	}
-	// wait until enough data has been collected, too many errors encountered,
-	// or the timeout is reached
-	err := sc.wait()
+	batchAdder := NewNmtNodeAdder(parentCtx, ipld.NewBatch(parentCtx, dag, ipld.MaxSizeBatchOption(batchSize(edsWidth))))
+	tree := wrapper.NewErasuredNamespacedMerkleTree(uint64(edsWidth)/2, nmt.NodeVisitor(batchAdder.Visit))
+	extended, err := rsmt2d.RepairExtendedDataSquare(dah.RowsRoots, dah.ColumnRoots, dataSquare, codec, tree.Constructor)
 	if err != nil {
 		return nil, err
 	}
-	// flatten the square
-	flattened := sc.flatten()
-	// repair the square
-	tree := wrapper.NewErasuredNamespacedMerkleTree(uint64(edsWidth) / 2)
-	return rsmt2d.RepairExtendedDataSquare(rowRoots, colRoots, flattened, codec, tree.Constructor)
-}
 
-// uniqueRandNumbers generates count unique random numbers with a max of max
-func uniqueRandNumbers(count, max int) []uint32 {
-	if count > max {
-		panic(fmt.Sprintf("cannot create %d unique samples from a max of %d", count, max))
-	}
-	samples := make(map[uint32]struct{}, count)
-	for i := 0; i < count; {
-		// nolint:gosec // G404: Use of weak random number generator
-		sample := uint32(rand.Intn(max))
-		if _, has := samples[sample]; has {
-			continue
-		}
-		samples[sample] = struct{}{}
-		i++
-	}
-	out := make([]uint32, count)
-	counter := 0
-	for s := range samples {
-		out[counter] = s
-		counter++
-	}
-	return out
+	return extended, batchAdder.Commit()
 }
 
-type index struct {
-	row uint32
-	col uint32
+type quadrant struct {
+	isLeftSubtree bool
+	from          int
+	chunk         [][]byte //roots
```

Comment:
Rename to Root. Also, I suggest having roots be always as CID's so you convert them in the beginning. This way it's also clear what roots are and not just arbitrary bytes.

Reply:
renamed to `rootCids`